File: app/src/main/java/com/faceunity/fulivedemo/ui/XfermodeRadioButton.java
package com.faceunity.fulivedemo.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.util.AttributeSet;
import android.util.TypedValue;

import com.faceunity.fulivedemo.R;

/**
 * Created by tujh on 2018/3/2.
 */

public class XfermodeRadioButton extends android.support.v7.widget.AppCompatRadioButton {

    private String textXfermode;
    private int textSizeXfermode;
    private int textColorXfermodeNormal = 0xffffffff;
    private int textColorXfermodeChecked = 0x99050f14;
    private int textXfermodeWidth;
    private int baseLineY;

    private Paint mMyRadioButtonPaint;

    public XfermodeRadioButton(Context context) {
        this(context, null);
    }

    public XfermodeRadioButton(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.radioButtonStyle);
    }

    public XfermodeRadioButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.xfermode_radio_btn, defStyleAttr, 0);

        textXfermode = typedArray.getString(R.styleable.xfermode_radio_btn_text_xfermode);
        textSizeXfermode = typedArray.getDimensionPixelSize(R.styleable.xfermode_radio_btn_text_size_xfermode, (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 12, context.getResources().getDisplayMetrics()));

        mMyRadioButtonPaint = new Paint();
        mMyRadioButtonPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.XOR));
        mMyRadioButtonPaint.setColor(textColorXfermodeNormal);
        mMyRadioButtonPaint.setTextSize(textSizeXfermode);
        mMyRadioButtonPaint.setAntiAlias(true);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        textXfermodeWidth = (int) mMyRadioButtonPaint.measureText(textXfermode);
        Paint.FontMetrics fontMetrics = mMyRadioButtonPaint.getFontMetrics();
        float top = fontMetrics.top;//为基线到字体上边框的距离,即上图中的top
        float bottom = fontMetrics.bottom;//为基线到字体下边框的距离,即上图中的bottom
        baseLineY = (int) (getMeasuredHeight() - top - bottom) / 2;
        setMeasuredDimension(40 + textXfermodeWidth, getMeasuredHeight());
    }

    @Override
    public void setChecked(boolean checked) {
        super.setChecked(checked);
        if (mMyRadioButtonPaint != null) {
            mMyRadioButtonPaint.setColor(checked ? textColorXfermodeChecked : textColorXfermodeNormal);
            mMyRadioButtonPaint.setXfermode(new PorterDuffXfermode(checked ? PorterDuff.Mode.SRC_ATOP : PorterDuff.Mode.XOR));
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawText(textXfermode, (getMeasuredWidth() - textXfermodeWidth) / 2, baseLineY, mMyRadioButtonPaint);
    }
}

File: fu_ui/src/main/java/com/faceunity/ui/radio/XfermodeRadioButton.kt
package com.faceunity.ui.radio

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.drawable.StateListDrawable
import android.util.AttributeSet
import android.util.TypedValue
import androidx.appcompat.widget.AppCompatRadioButton
import com.faceunity.ui.R

/**
 * Created by tujh on 2018/3/2.
 */
class XfermodeRadioButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.radioButtonStyle
) : AppCompatRadioButton(context, attrs, defStyleAttr) {
    private val textXfermode: String?
    private val textSizeXfermode: Int
    private val textColorXfermodeNormal = -0x1
    private val textColorXfermodeChecked = -0x66faf0ec
    private var textXfermodeWidth = 0
    private var baseLineY = 0
    private val mMyRadioButtonPaint: Paint?
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        textXfermodeWidth = mMyRadioButtonPaint!!.measureText(textXfermode).toInt()
        val fontMetrics = mMyRadioButtonPaint.fontMetrics
        val top = fontMetrics.top //为基线到字体上边框的距离,即上图中的top
        val bottom = fontMetrics.bottom //为基线到字体下边框的距离,即上图中的bottom
        baseLineY = (measuredHeight - top - bottom).toInt() / 2
        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    override fun setChecked(checked: Boolean) {
        super.setChecked(checked)
        if (mMyRadioButtonPaint != null) {
            mMyRadioButtonPaint.color = if (checked) textColorXfermodeChecked else textColorXfermodeNormal
            mMyRadioButtonPaint.xfermode = PorterDuffXfermode(if (checked) PorterDuff.Mode.SRC_ATOP else PorterDuff.Mode.XOR)
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val x = (measuredWidth - textXfermodeWidth) / 2
        canvas.drawText(textXfermode!!, x.toFloat(), baseLineY.toFloat(), mMyRadioButtonPaint!!)
    }

    init {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.xfermode_radio_btn, defStyleAttr, 0)
        textXfermode = typedArray.getString(R.styleable.xfermode_radio_btn_text_xfermode)
        textSizeXfermode = typedArray.getDimensionPixelSize(
            R.styleable.xfermode_radio_btn_text_size_xfermode,
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 12f, context.resources.displayMetrics).toInt()
        )
        typedArray.recycle()
        mMyRadioButtonPaint = Paint()
        mMyRadioButtonPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.XOR)
        mMyRadioButtonPaint.color = textColorXfermodeNormal
        mMyRadioButtonPaint.textSize = textSizeXfermode.toFloat()
        mMyRadioButtonPaint.isAntiAlias = true
        buttonDrawable = StateListDrawable()
    }
}